File: 0890-find-and-replace-pattern/0890-find-and-replace-pattern.java
class Solution {
public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> list = new ArrayList<String>();
            for(String word: words){
             if(isPatternMatch(word, pattern))
                 list.add(word);
         }
        return list;
    }
    
    
    private boolean isPatternMatch(String word, String pattern){
      HashMap<Character, Character> map = new HashMap<>();
      if(word.length()!=pattern.length()) return false;
      for(int i=0;i<pattern.length();i++){
            if(map.containsKey(pattern.charAt(i))){
             if (map.get(pattern.charAt(i))!=word.charAt(i)) return false;
             }  else if(map.containsValue(word.charAt(i))) return false;
            map.put(pattern.charAt(i), word.charAt(i));
        }
         return true;
    }
}

File: 0890-find-and-replace-pattern/0890-find-and-replace-pattern.py
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        def match(words):
            d1,d2 = {},{}
            for w,p in zip(words,pattern):
                if w not in d1: 
                    d1[w] = p
                if p not in d2: 
                    d2[p] = w
                if (d1[w], d2[p]) != (p, w):
                    return False
            return True
        
        return filter(match, words)